### PR TITLE
Cas 285/attachment validation

### PR DESCRIPTION
**Pull request by devUndef1ned**
According to [Jira ticket ](https://stream-io.atlassian.net/browse/CAS-285) we have to provide functionality to check that an attachment has a valid url link
1) Added AttachmentHelper with hasValidUrl method
2) Added tests
3) Update mockito version

### CODE REVIEW

File: client/build.gradle [hunk 1]
```diff
@@ -72,17 +72,21 @@ dependencies {
 
     implementation 'com.github.mrmike:ok2curl:0.6.0'
 
+    implementation 'com.facebook.stetho:stetho-okhttp3:1.5.1'
```
Comment (zsmb13) [inline client/build.gradle]: I think we could make this be `debugImplementation` only, with [something like this](https://stackoverflow.com/a/31483962/4465208). Just to make sure we're not including it in release versions.

File: client/src/main/java/io/getstream/chat/android/client/ChatClient.kt [hunk 2]
```diff
@@ -305,6 +308,8 @@ interface ChatClient {
 
     fun getVersion(): String
 
+    fun hasValidUrl(attachment: Attachment): Boolean
```
Comment (zsmb13) [inline client/src/main/java/io/getstream/chat/android/client/ChatClient.kt]: Where is this method going to be called from? Does it have to exposed through `ChatClient` specifically?

File: client/src/main/java/io/getstream/chat/android/client/helpers/AttachmentHelper.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package io.getstream.chat.android.client.helpers
+
+import io.getstream.chat.android.client.models.Attachment
+import io.getstream.chat.android.client.utils.SystemTimeProvider
+import org.apache.commons.validator.routines.UrlValidator
+
+class AttachmentHelper(private val systemTimeProvider: SystemTimeProvider = SystemTimeProvider()) {
```
Comment (zsmb13) [inline client/src/main/java/io/getstream/chat/android/client/helpers/AttachmentHelper.kt:7]: Can this class be internal? We already have its functionality exposed through `ChatClient`.

File: client/src/test/java/io/getstream/chat/android/client/api/ClientConnectionTests.kt [hunk 4]
```diff
@@ -77,6 +79,7 @@ internal class ClientConnectionTests {
         logger = mock(ChatLogger::class.java)
         notificationsManager = mock(ChatNotifications::class.java)
         initConnectionListener = mock(InitConnectionListener::class.java)
+        attachmentHelper = mock(AttachmentHelper::class.java)
```
Comment (zsmb13) [inline client/src/test/java/io/getstream/chat/android/client/api/ClientConnectionTests.kt]: Let's update these `mock` calls to use the reified function instead of using a `Class` parameter too, since we've done it in `MockClientBuilder` too 🙂

File: client/src/main/java/io/getstream/chat/android/client/helpers/AttachmentHelper.kt [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+            urlValidator.isValid(url).not() -> false
+            url.contains(KEY_WORD_EXPIRES) -> {
+                parseTimeStampOrNull(url)?.let { timestamp -> timestamp > systemTimeProvider.provideTime() } ?: false
+            }
+            else -> true
+        }
+    }
+
+    private fun parseTimeStampOrNull(url: String): Long? {
+        return url.substringAfter(KEY_WORD_EXPIRES).substringBefore(CHAR_AMPERSAND).toLongOrNull()
```
Comment (zsmb13) [inline client/src/main/java/io/getstream/chat/android/client/helpers/AttachmentHelper.kt]: We shouldn't do string manipulation for URL parsing, we can do this through utilities, it's kind of dangerous (makes too many assumptions about the structure of the URL).

Creating an [`android.net.Uri`](https://developer.android.com/reference/android/net/Uri) with `Uri.parse(url)` and then using methods like [`getQueryParameter`](https://developer.android.com/reference/android/net/Uri#getQueryParameter(java.lang.String)) can help us check if we have the query param is present, and what its value is.

**Review by zsmb13 — CHANGES_REQUESTED** [review]

File: client/src/main/java/io/getstream/chat/android/client/helpers/AttachmentHelper.kt (hunk 3 above)
Comment (JcMinarro) [reply client/src/main/java/io/getstream/chat/android/client/helpers/AttachmentHelper.kt:7]: I think it can expose this class and remove `ChatClient::hasValidUrl()` method.
We don't need it there

File: client/src/main/java/io/getstream/chat/android/client/ChatClient.kt [hunk 6]
```diff
@@ -394,6 +399,10 @@ interface ChatClient {
                 throw IllegalStateException("apiKey is not defined in " + this::class.java.simpleName)
             }
 
+            if (BuildConfig.DEBUG) {
+                Stetho.initializeWithDefaults(appContext)
```
Comment (JcMinarro) [inline client/src/main/java/io/getstream/chat/android/client/ChatClient.kt]: Could we keep Stetho installation into another different PR?

**Review by JcMinarro — CHANGES_REQUESTED** [review]

File: client/build.gradle (hunk 1 above)
Comment (devUndef1ned) [reply client/build.gradle]: Remove Stetho from this PR

File: client/src/main/java/io/getstream/chat/android/client/ChatClient.kt (hunk 2 above)
Comment (devUndef1ned) [reply client/src/main/java/io/getstream/chat/android/client/ChatClient.kt]: Remove from here, expose this functionality via Helper

File: client/src/main/java/io/getstream/chat/android/client/helpers/AttachmentHelper.kt (hunk 5 above)
Comment (devUndef1ned) [reply client/src/main/java/io/getstream/chat/android/client/helpers/AttachmentHelper.kt]: Thanks for point! Replaced this logic by logic of okhttp library.

File: client/src/test/java/io/getstream/chat/android/client/api/ClientConnectionTests.kt (hunk 4 above)
Comment (devUndef1ned) [reply client/src/test/java/io/getstream/chat/android/client/api/ClientConnectionTests.kt]: Done

File: client/src/main/java/io/getstream/chat/android/client/ChatClient.kt (hunk 6 above)
Comment (devUndef1ned) [reply client/src/main/java/io/getstream/chat/android/client/ChatClient.kt]: Yes, removed it from here

File: client/src/main/java/io/getstream/chat/android/client/helpers/AttachmentHelper.kt [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+            url.contains(QUERY_KEY_NAME_EXPIRES) -> {
+                parseTimeStampOrNull(url)?.let { timestamp -> timestamp > systemTimeProvider.provideTime() } ?: false
+            }
+            else -> true
+        }
+    }
+
+    private fun parseTimeStampOrNull(url: String): Long? {
+        return url.toHttpUrlOrNull()?.queryParameter(QUERY_KEY_NAME_EXPIRES)?.toLongOrNull()
+    }
```
Comment (zsmb13) [inline client/src/main/java/io/getstream/chat/android/client/helpers/AttachmentHelper.kt]: Since we started using `toHttpUrlOrNull` from OkHttp (nice, didn't know about this!), which as I understand validates the URL while parsing, we could simplify all the way to something like this:

```suggestion
    fun hasValidUrl(attachment: Attachment): Boolean {
        val timeStamp = attachment.url
            ?.toHttpUrlOrNull() 
            ?.queryParameter(QUERY_KEY_NAME_EXPIRES)
            ?.toLongOrNull()
            ?: return false
        return timeStamp > systemTimeProvider.provideTime()
    }
```

This would also let us drop the URL validator dependency.

Comment (devUndef1ned) [reply client/src/main/java/io/getstream/chat/android/client/helpers/AttachmentHelper.kt]: Your code seems very elegant, but unfortunately it fails test when url doesn't contain Query param "expires" (I'm not sure, maybe such url without expires should not be valid). But you're right we can omit validator dependency

**Review by JcMinarro — DISMISSED** [review]
LGTM

File: client/src/main/java/io/getstream/chat/android/client/helpers/AttachmentHelper.kt (hunk 7 above)
Comment (zsmb13) [reply client/src/main/java/io/getstream/chat/android/client/helpers/AttachmentHelper.kt]: Oops, that's correct, I missed that case when writing this up. Nicely done with the tests that they caught it 👍

File: client/src/main/java/io/getstream/chat/android/client/helpers/AttachmentHelper.kt [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+import io.getstream.chat.android.client.models.Attachment
+import io.getstream.chat.android.client.utils.SystemTimeProvider
+import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
+
+class AttachmentHelper(private val systemTimeProvider: SystemTimeProvider = SystemTimeProvider()) {
+
+    fun hasValidUrl(attachment: Attachment): Boolean {
+        val url = attachment.url?.toHttpUrlOrNull() ?: return false
+        if (url.queryParameterNames.contains(QUERY_KEY_NAME_EXPIRES).not()) {
```
Comment (zsmb13) [inline client/src/main/java/io/getstream/chat/android/client/helpers/AttachmentHelper.kt:11]: Could also be expressed like this, which you might find more readable:

```suggestion
        if (QUERY_KEY_NAME_EXPIRES !in url.queryParameterNames) {
```

File: client/src/main/java/io/getstream/chat/android/client/helpers/AttachmentHelper.kt [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+            return true
+        }
+        return url.queryParameter(QUERY_KEY_NAME_EXPIRES)?.toLongOrNull()
+            ?.let { timestamp -> timestamp > systemTimeProvider.provideTime() } ?: false
```
Comment (zsmb13) [inline client/src/main/java/io/getstream/chat/android/client/helpers/AttachmentHelper.kt]: I believe this code is correct here, but it's the kind of suspicious structure with `let` and the Elvis operator that can occasionally lead to unexpected behaviour. It took me a while to think through all the possible cases of what can be `null` at what point, and what that causes.

How about going with something more conventional? (It's the same amount of code anyway, but much easier to verify what it does imo)

```suggestion
        val timestamp = url.queryParameter(QUERY_KEY_NAME_EXPIRES)?.toLongOrNull() ?: return false
        return timestamp > systemTimeProvider.provideTime()
```

**Review by zsmb13 — DISMISSED** [review]
Have two more suggestions for code style, but looks correct otherwise 👍

File: client/src/main/java/io/getstream/chat/android/client/helpers/AttachmentHelper.kt (hunk 8 above)
Comment (devUndef1ned) [reply client/src/main/java/io/getstream/chat/android/client/helpers/AttachmentHelper.kt:11]: Don't see big difference here. Just matter of taste. Thanks anyway

**Review by zsmb13 — APPROVED** [review]
